wagtail/users/views/users.py

```diff
@@ -70,74 +70,70 @@ def get_users_filter_query(q, model_fields):
     return conditions
 
 
-@any_permission_required(add_user_perm, change_user_perm, delete_user_perm)
-@vary_on_headers("X-Requested-With")
-def index(request, *args):
-    q = None
-    is_searching = False
-
-    group = None
-    group_filter = Q()
-    if args:
-        group = get_object_or_404(Group, id=args[0])
-        group_filter = Q(groups=group) if args else Q()
+class Index(IndexView):
+
+    any_permission_required = ["add", "change", "delete"]
+    permission_policy = ModelPermissionPolicy(User)
+    model = User
+    context_object_name = "users"
+    index_url_name = "wagtailusers_users:index"
+    add_url_name = "wagtailusers_users:add"
+    edit_url_name = "wagtailusers_users:edit"
+    default_ordering = "name"
+    paginate_by = 20
+    template_name = None
+    is_searchable = True
+    page_title = gettext_lazy("Users")
+
+    def setup(self, request, *args, **kwargs):
+        super().setup(request, *args, **kwargs)
+        setattr(self, "template_name", self.get_template())
+        self.group = get_object_or_404(Group, id=args[0]) if args else None
+        self.group_filter = Q(groups=self.group) if self.group else Q()
+        self.model_fields = [f.name for f in User._meta.get_fields()]
+
+    def get_queryset(self):
+        if self.is_searching:
+            conditions = get_users_filter_query(self.search_query, self.model_fields)
+            users = User.objects.filter(self.group_filter & conditions)
+        else:
+            users = User.objects.filter(self.group_filter)
 
-    model_fields = [f.name for f in User._meta.get_fields()]
+        if self.locale:
+            users = users.filter(locale=self.locale)
 
-    if "q" in request.GET:
-        form = SearchForm(request.GET, placeholder=_("Search users"))
-        if form.is_valid():
-            q = form.cleaned_data["q"]
-            is_searching = True
-            conditions = get_users_filter_query(q, model_fields)
+        if "last_name" in self.model_fields and "first_name" in self.model_fields:
+            users = users.order_by("last_name", "first_name")
 
-            users = User.objects.filter(group_filter & conditions)
-    else:
-        form = SearchForm(placeholder=_("Search users"))
+        if self.get_ordering() == "username":
+            users = users.order_by(User.USERNAME_FIELD)
 
-    if not is_searching:
-        users = User.objects.filter(group_filter)
+        return users
 
-    if "last_name" in model_fields and "first_name" in model_fields:
-        users = users.order_by("last_name", "first_name")
+    @property
+    def requested_with_XMLHttpRequest(self):
```

Comment:
I understand the reasoning here, but let's go with a bit more of a verbose option that keeps things simple and aligns with existing conventions (just checking the request header where needed).

Longer term - this approach may be refactored to something like this property approach but for now it is not required.

Reply:
In other places, such as the choosers, we've phased out the use of `x-requested-with` in favour of having separate views at distinct URLs such as `/admin/users/` and `/admin/users/results/`, for the full page response versus the HTML fragment. (Of course, class-based views make this really easy, since they can share the same base class and just override the `template_name` property.)

I'll leave it up to you to decide whether you want to do that as part of this PR, or leave it for a future refactor - but either way, I agree that we shouldn't engineer things too strongly around the x-requested-with approach.

Reply:
Got it!

Reply:
@mehrdadmoradii  thanks for reviewing this, thanks Matt for the insights into the longer term plans here.

Mehrdad - I think it's probably simplest to focus this PR on the class based view change, this way we can build on this for the added URL path approach.

I will aim to create a future issue that documents this approach we want to take long term, or find one if it exists. The `x-requested-with` request header is part of the jQuery async request code so that will also be replaced over time.

Reply:
@lb- thanks for your review. Please ping me on that issue when created. 

Also, is there any other work that needs to get done?

Reply:
@gasman  - https://github.com/wagtail/wagtail/issues/9272 raised - please advise if this captures your thinking adequately.

@mehrdadmoradii  - no rush on this - but have a read and see if it makes sense for any similar future clean up of view code.

Reply:
@lb- #9272 looks good to me!